x/concentrated-liquidity/pool.go

```
@@ -75,128 +77,185 @@ func (p Pool) SpotPrice(ctx sdk.Context, baseAssetDenom string, quoteAssetDenom
+	// TODO: How do we remove/generalize this? I am stumped.
 	amountETH := int64(1000000)
 	amountUSDC := int64(5000000000)
 
-	liq0 := liquidity0(amountETH, sqrtPCurTick, sqrtPUpperTick)
-	liq1 := liquidity1(amountUSDC, sqrtPCurTick, sqrtPLowerTick)
+	// find liquidity of assetA and assetB
+	liq0 := liquidity0(amountETH, curSqrtPrice, sqrtPUpperTick)
+	liq1 := liquidity1(amountUSDC, curSqrtPrice, sqrtPLowerTick)
 
+	// utilize the smaller liquidity between assetA and assetB when performing the swap calculation
 	liq := sdk.MinDec(liq0, liq1)
-	if tokenIn.Denom == asset1 {
-		priceDiff := tokenAmountInAfterFee.Quo(liq)
-		priceNext := sqrtPCurTick.Add(priceDiff)
-		// new amount in, will be needed later
-		//amountIn = calcAmount1(liq, priceNext, sqrtPCurTick)
-		amountOut := calcAmount0Delta(liq, priceNext, sqrtPCurTick)
-		return sdk.NewCoin(tokenOutDenom, amountOut.TruncateInt()), nil
-	} else if tokenIn.Denom == asset0 {
-		priceNextTop := liq.Mul(sqrtPCurTick)
-		priceNextBot := liq.Add(tokenAmountInAfterFee.Mul(sqrtPCurTick))
-		priceNext := priceNextTop.Quo(priceNextBot)
-		// new amount in, will be needed later
-		//amountIn = calcAmount1(liq, priceNext, sqrtPCurTick)
-		amountOut := calcAmount1Delta(liq, priceNext, sqrtPCurTick)
-		return sdk.NewCoin(tokenOutDenom, amountOut.TruncateInt()), nil
-	}
-	return sdk.Coin{}, fmt.Errorf("tokenIn (%s) does not match any asset in pool", tokenIn.Denom)
+
+	swapState := SwapState{
+		amountSpecifiedRemaining: tokenAmountInAfterFee,
+		amountCalculated:         sdk.ZeroDec(),
+		sqrtPrice:                curSqrtPrice,
+		tick:                     priceToTick(curSqrtPrice.Power(2)),
+	}
+
+	// TODO: This should be GT 0 but some instances have very small remainder
+	// need to look into fixing this
+	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 6)) {
+		lte := tokenIn.Denom == asset1
+		nextTick, _ := k.NextInitializedTick(ctx, poolId, swapState.tick.Int64(), lte)
```

This could return `false` if there is no tick found. maybe return an error in that case?